### PR TITLE
Remove `unsafe_type_of` from `Coercion`

### PR DESCRIPTION
**Pull request by maximedenes**
We thread explicitly the evar map everywhere for this purpose.

https://github.com/coq-community/paramcoq/pull/48

### CODE REVIEW

File: pretyping/coercion.ml [hunk 1]
```diff
@@ -254,115 +260,114 @@ and coerce ?loc env evdref (x : EConstr.constr) (y : EConstr.constr)
...
-	   | _ -> subco ())
-      | _, _ ->  subco ()
+                 let lam_type = Typing.unsafe_type_of env sigma c in
+                 let lam_type' = Typing.unsafe_type_of env sigma c' in
+                   sigma, coerce_application sigma lam_type lam_type' c c' l l'
+               else subco sigma
+           | _ -> subco sigma)
+      | _, _ ->  subco sigma
 
   and subset_coerce env evdref x y =
```
Comment (SkySkimmer) [inline pretyping/coercion.ml]: Missed evdref

Comment (maximedenes) [reply pretyping/coercion.ml]: Well spotted, thanks!

File: pretyping/coercion.ml [hunk 2]
```diff
@@ -253,115 +258,110 @@ let coerce ?loc env evdref (x : EConstr.constr) (y : EConstr.constr)
...
-                 let lam_type' = Typing.unsafe_type_of env evm c' in
-                   coerce_application lam_type lam_type' c c' l l'
-               else subco ()
-           | _ -> subco ())
-      | _, _ ->  subco ()
-
-  and subset_coerce env evdref x y =
-    match disc_subset !evdref x with
+                 let lam_type = Typing.unsafe_type_of env sigma c in
+                 let lam_type' = Typing.unsafe_type_of env sigma c' in
```
Comment (SkySkimmer) [inline pretyping/coercion.ml]: Remaining unsafe

**Comment by SkySkimmer** [comment]
Rebased past coercion traces

File: pretyping/coercion.ml [hunk 3]
```diff
@@ -136,70 +140,71 @@ let lift_args n sign =
...
+            let args = List.rev (restargs @ mkRel 1 :: List.map (lift 1) tele) in
+            let pred = mkLambda (n, eqT, applist (lift 1 c, args)) in
+            let sigma, eq = papp sigma coq_eq_ind [| eqT; hdx; hdy |] in
+            let sigma, evar = make_existential ?loc n.binder_name env sigma eq in
+            let eq_app sigma x = papp sigma coq_eq_rect
+                [| eqT; hdx; pred; x; hdy; evar|]
+            in
+            aux sigma (hdy :: tele) (subst1 hdx restT)
+              (subst1 hdy restT') (succ i)  (fun sigma x -> let sigma, x = co sigma x in eq_app sigma x)
+        else Some (fun sigma x ->
```
Comment (SkySkimmer) [inline pretyping/coercion.ml]: Here we drop the current sigma
Let's see what happens once that's fixed

File: pretyping/coercion.ml [hunk 4]
```diff
@@ -552,45 +550,46 @@ let rec inh_conv_coerce_to_fail ?loc env evd ?(flags=default_flags_of env) rigid
...
-            inh_conv_coerce_to_fail ?loc env1 evd rigidonly
+          let (sigma, v1, trace1) =
+            inh_conv_coerce_to_fail ?loc env1 sigma rigidonly
               (mkRel 1) (lift 1 u1) (lift 1 t1) in
-          let v2 = beta_applist evd' (lift 1 v,[v1]) in
-          let t2 = Retyping.get_type_of env1 evd' v2 in
-          let (evd'',v2',trace2) = inh_conv_coerce_to_fail ?loc env1 evd' rigidonly v2 t2 u2 in
+          let v2 = beta_applist sigma (lift 1 v,[v1]) in
+          let t2 = Retyping.get_type_of env1 sigma v2 in
+          let (sigma,v2',trace2) = inh_conv_coerce_to_fail ?loc env1 sigma rigidonly v2 t2 u2 in
```
Comment (SkySkimmer) [inline pretyping/coercion.ml]: BTW I removed all `evd` to `sigma` and removed `'` to make sure we didn't miss a shadowing somewhere.

File: pretyping/coercion.ml (hunk 2 above)
Comment (SkySkimmer) [reply pretyping/coercion.ml]: Removed them

**Comment by SkySkimmer** [comment]
Going to need an overlay for paramcoq, I guess they use `papp`

File: pretyping/coercion.ml [hunk 5]
```diff
@@ -136,222 +140,219 @@ let lift_args n sign =
...
+            with NoSubtacCoercion ->
+              let sigma, typ = Typing.type_of env sigma c in
+              let sigma, typ' = Typing.type_of env sigma c' in
+              coerce_application sigma typ typ' c c' l l')
+         else
+           subco sigma
+       | x, y when EConstr.eq_constr sigma c c' ->
+         if Int.equal (Array.length l) (Array.length l') then
+           let sigma, lam_type = Typing.type_of env sigma c in
+           let sigma, lam_type' = Typing.type_of env sigma c' in
```
Comment (mattam82) [inline pretyping/coercion.ml]: I think `Retyping.get_type_of` should be enough, these are already typed terms.

**Review by mattam82 — APPROVED** [review]
The PR looks fine to me overall, giving less chances to mishandle the evar_map state. Consistent naming is nice too :)

**Comment by ejgallego** [comment]
This should enable to mark `unsafe_type_of` as deprecated after #11538

**Comment by maximedenes** [comment]
Overlay added.

**Comment by SkySkimmer** [comment]
You must link the overlay PR in the PR opening message.

**Comment by maximedenes** [comment]
@SkySkimmer I opened the paramcoq PR and added the link above.

**Review by SkySkimmer — APPROVED** [review]
I'm going to assume nobody started using papp since last CI and merge.

**Pull request merged**